crates/dojo/core/src/meta/introspect.cairo

```diff
@@ -1,6 +1,151 @@
+        false,
+    ],
+    // u16
+    [
+        false, false, true, true, true, true, false, false, false, false, false, false, true, false,
+        false,
+    ],
+    // u32
+    [
+        false, false, false, true, true, true, false, false, false, false, false, false, true,
+        false, false,
+    ],
+    // u64
+    [
+        false, false, false, false, true, true, false, false, false, false, false, false, true,
+        false, false,
+    ],
+    // u128
+    [
+        false, false, false, false, false, true, false, false, false, false, false, false, true,
+        false, false,
+    ],
+    // u256
+    [
+        false, false, false, false, false, false, true, false, false, false, false, false, false,
+        false, false,
+    ],
+    // i8
+    [
+        false, false, false, false, false, false, false, true, true, true, true, true, true, false,
+        false,
+    ],
+    // i16
+    [
+        false, false, false, false, false, false, false, false, true, true, true, true, true, false,
+        false,
+    ],
+    // i32
+    [
+        false, false, false, false, false, false, false, false, false, true, true, true, true,
+        false, false,
+    ],
+    // i64
+    [
+        false, false, false, false, false, false, false, false, false, false, true, true, true,
+        false, false,
+    ],
+    // i128
+    [
+        false, false, false, false, false, false, false, false, false, false, false, true, true,
+        false, false,
+    ],
+    // felt252
+    [
+        false, false, false, false, false, false, false, false, false, false, false, false, true,
+        true, true,
+    ],
+    // ClassHash
+    [
+        false, false, false, false, false, false, false, false, false, false, false, false, true,
+        true, true,
+    ],
+    // ContractAddress
+    [
+        false, false, false, false, false, false, false, false, false, false, false, false, true,
+        true, true,
+    ],
+];
+
+#[inline(always)]
+fn primitive_to_index(primitive: felt252) -> u32 {
+    if primitive == 'bool' {
+        return 0;
+    }
+    if primitive == 'u8' {
+        return 1;
+    }
+    if primitive == 'u16' {
+        return 2;
+    }
+    if primitive == 'u32' || primitive == 'usize' {
+        return 3;
+    }
+    if primitive == 'u64' {
+        return 4;
+    }
+    if primitive == 'u128' {
+        return 5;
+    }
+    if primitive == 'u256' {
+        return 6;
+    }
+    if primitive == 'i8' {
+        return 7;
+    }
+    if primitive == 'i16' {
+        return 8;
+    }
+    if primitive == 'i32' {
+        return 9;
+    }
+    if primitive == 'i64' {
+        return 10;
+    }
+    if primitive == 'i128' {
+        return 11;
+    }
+    if primitive == 'felt252' {
+        return 12;
+    }
+    if primitive == 'ClassHash' {
+        return 13;
+    }
+    if primitive == 'ContractAddress' {
+        return 14;
+    }
+
+    return 0xFFFFFFFF;
```

Comment:
Let's use a constant here, that can be re-use to test the returned value before actually using it as an index in the array.

Or, we could have the function to panic, if a primitive is not supported. May even be better at some point, since we're supposed to cover them?

Reply:
Ah yes, I forgot to handle this case (I've put a random value temporarily ...), sorry.

Yes, I think we can just panic here as we should never reach this point.
We can also check if the primitive type is supported when we process the `Introspect` derive attribute in the `lang` crate. So we will get a compilation error instead of a runtime error. What do you think ?

Reply:
Yeah good point if we can catch something at compile time, even better. But it will only be that an unsupported type (which shouldn't be much right?). Since the comparison for validating the upgrade is only runtime, correct?

Reply:
Yes, the idea would be to check if the given primitive is in a supported primitive set (so the 16 primitive types used in upgrade validation).
It would directly tell to the user that a primitive is not supported at compile time.
So yes, it will catch a few existing types that we don't support yet such as maybe `bytes31` or `Nullable<T>` for example, and new future types like `u512` if it appears one day.

Reply:
In fact, if a not supported primitive type is used, we will already get a compilation error due to the fact that there is no implementation of `Introspect` for this type.
So, I don't know if we let like this or if we add a check to provide a more explicit error message 🤔

Reply:
It's good for now how you've added the panic. 👍 The idea is having as much compile time error as we can.